### DISCUSSION TITLE
Contract to Contract Calls on Algorand

### DISCUSSION THREAD

**Discussion by BunsanMuchi** [reach-sh/reach-lang#1443] [Q&A] [answered]
If I'm doing a c2c call using an ABI that requires a deposit (for example calling HumbleSwap or Pact's swap), I can import this ABI as remote object.

But to interact with it, do I have to feed it the Contracts ID, get the ABI and then use the .pay() method for the remote object, or do I have to use the contract address?

It would also help me a lot to get a quick overview for c2c calls since I do know they can get a bit tricky with the fee structure and local-state quirks Algorand has.

**Comment by jeapostrophe** [comment, marked as answer]
I don't understand what you mean by "but to interact with it"... a remote object IS for interacting with it.

When you make a remote object, you create it with the Contract value (that's the id on Algorand) and the interface.

When you call a function, you access the fields of the remote object, add any augmentations, then call it. 

I am stupid. I actually do not know any way to plain it other than what is written here https://docs.reach.sh/rsh/consensus/#remote-objects

There are multiple code samples in that section.

Is there an api you are having trouble calling?

**Reply by BunsanMuchi** [reply]
That first sentence makes more sense to me, I think I got it running like I wanted now, but I still don't know when to use "Opt-ins" and when not....

As for the the linked part of the docs, it's not a problem with the what's written in the docs. In fact, you're right, it's extremely plainly written and simple to understand. However, I just don't understand HOW to apply some aspects of the ALGO-specific implementation (if that makes any sense).  Usually when I read the reach docs I see some snippets that are like gold for me since they tend to be exactly the small tidbits that make my life incredibly easier, and for some reason on one of the most complex aspects I've encountered so far (dealing with ALGO-remote objects) I didn't see (unless by my mistake) any.

Take for example something like the fee aspect, it took me until a private conversation with you to find out that there was a fee detailed when running the DEBUG mode, I don't think that there's somewhere on the docs that details that. Or something like the "Opt-in" feature, to this moment that one keeps on being tricky for me to fully test out/wrap my head around.

I understand that these are TEAL-specific things that are in and of itself hard to  detail without opening a whole can of worms, but I really don't see how it can be avoided. If anything, it's the fact that Reach has spoiled  me so much with abstracting away a lot, that this aspect hit me by surprise and I found harder to get right than other things in my development experience with the language.

**Reply by jeapostrophe** [reply]
I think I understand the problem. We are actively working on a new Reach feature where you can import one Reach.App into another so that all of these settings are determined for you. In the mean time, I am very happy to answer any specific questions like "How can I call _this_ function? I get _this_ error." The answer will basically always be "Add `.ALGO({ whatever: xyz })`"

**Reply by BunsanMuchi** [reply]
That would help A LOT, since at the very least I get more clues as to what went wrong and I can look there. Like always, thanks for all the help!

**Reply by jeapostrophe** [reply]
Cool, send 'em over
